edb/server/protocol/pg_ext.pyx

```diff
@@ -0,0 +1,763 @@
+                            f"cursor \"{name}\" does not exist",
+                        )
+                    # No portal lives outside transactions, so we just clear
+                    # the replay cache here
+                    del self.portals[name]
+
+                else:
+                    raise pgerror.ProtocolViolation("invalid Close kind")
+
+            except Exception as ex:
+                self.write_error(ex)
+                self.flush()
+                self.ignore_till_sync = True
+
+            else:
+                buf = WriteBuffer.new_message(b'3')  # CloseComplete
+                self.write(buf.end_message())
+
+        elif mtype == b'H':  # Flush
+            self.buffer.finish_message()
+            if self.debug:
+                self.debug_print("Flush")
+            self.flush()
+
+        else:
+            if self.debug:
+                self.debug_print(
+                    "MESSAGE", chr(mtype), self.buffer.consume_message()
+                )
+            raise pgerror.FeatureNotSupported()
+
+    async def extended_query(self):
+        cdef:
+            WriteBuffer buf
+            int16_t i
+            bytes data
+
+        actions = deque()
+        fresh_stmts = set()
+        fresh_portals = set()
+        executed_portals = set()
+        suspended_portals = set()
+
+        # Here we will exhaust the buffer and queue up actions for the backend.
+        # Any error in this step will be handled in the outer main_step() -
+        # the error will be returned, any remaining messages in the buffer will
+        # be discarded until a Sync message is found (ignore_till_sync).
+        # This also means no partial action is executed in the backend for now.
+        while self.buffer.take_message():
+            mtype = self.buffer.get_message_type()
+
+            if mtype == b'P':  # Parse
+                stmt_name = self.buffer.read_null_str().decode(
+                    self.client_encoding
+                )
+                query_bytes = self.buffer.read_null_str()
+                query_str = query_bytes.decode(self.client_encoding)
+                data = self.buffer.consume_message()
+                if self.debug:
+                    self.debug_print("Parse", repr(stmt_name), query_str, data)
+
+                if stmt_name and stmt_name in self.prepared_stmts:
+                    raise pgerror.new(
+                        pgerror.ERROR_DUPLICATE_PREPARED_STATEMENT,
+                        f"prepared statement \"{stmt_name}\" already exists",
+                    )
+
+                sql_source = await self.compile(query_str)
+                if len(sql_source) > 1:
+                    raise pgerror.new(
+                        pgerror.ERROR_SYNTAX_ERROR,
+                        "cannot insert multiple commands into a prepared "
+                        "statement",
+                    )
+                sql_text = sql_source[0].encode("utf-8")
```

Comment:
Result of parsing + resolving may also be SQL 0 statements - `sql_source[0]` will throw an error.